okhttp/src/main/java/com/squareup/okhttp/IdnaMappingTable.java

```diff
@@ -0,0 +1,394 @@
+ *   <li>Ignored code points. For example, U+FEFF is discarded if present, so "square\ufeff.com."
+ *       yields "square.com".
+ *   <li>Mapped code points. This includes case mapping, like "SQUARE.COM" to "square.com" and
+ *       "Σ.com" to "σ.com". Plus more exotic mappings such as for ℡ — the telephone symbol — which
+ *       maps to three code points "tel".
+ *   <li>Disallowed code points. The cyrillic letter palochka "Ӏ" is disallowed, and mapping
+ *       hostnames that contain it will return null.
+ * </ul>
+ *
+ * <h3>Implementation Notes</h3>
+ *
+ * <p>The IDNA table is specified as a 768 KiB human-readable text file. In the OkHttp .jar file we
+ * ship a smaller 46 KIB machine-readable binary file. The two formats are functionally equivalent,
+ * and this class has APIs to read both.
+ *
+ * <p>This class attempts to balance memory and runtime efficiency, both when it reads the IDNA
+ * binary file and when using the resulting tables to map code points. Data is modeled as ranges of
+ * code points: a range like [U+0061..U+007A] is stored as a single element in each of two unsigned
+ * short arrays. To look up the value for U+0062, we binary search to find the entry for U+0061, and
+ * look up the value shared for that entire range.
+ *
+ * <p>Since code points are 21-bit numbers (0x0..10ffff), we split them into pages that each contain
+ * a 16 bit range. For example, values in 0x30000..0x3ffff are in page 0x3 which is then binary
+ * searched over a 16 bit suffix: between 0x0000..0xffff. We save ~16 KiB of memory using unsigned
+ * shorts instead of ints as the subjects of our binary search.
+ *
+ * <p>The value of each range is also an unsigned short, containing a status (valid, ignored,
+ * mapped, etc.) and either an IDNA2008 status (NV8 or XV8) or an offset to look up the replacement
+ * string. All replacement strings are concatenated substrings of a shared ~8,600 character string.
+ *
+ * <p>This class uses bit packing to store several values in a single unsigned short. There are two
+ * bit packing choices:
+ * <ul>
+ *   <li>Ranges with status "mapped" or "disallowed_STD3_mapped" use 2 bits for the bit packing type
+ *       and status, plus 14 bits for the offset of the mapped string.
+ *   <li>Other ranges use 2 bits for the bit packing type, 6 bits that are unused, 4 bits for IDNA
+ *       2008 status and 4 bits for status.
+ * </ul>
+ *
+ * <p>This documentation refers to "unsigned short". Java's unsigned short type is named 'char',
+ * but none of the arrays in this class contain UTF-16 code points and should not be interpreted
+ * as text.
+ */
+final class IdnaMappingTable {
+  private static final BitPackingType[] BIT_PACKING_TYPE_VALUES = BitPackingType.values();
+  private static final Status[] STATUS_VALUES = Status.values();
+  private static final Idna2008Status[] IDNA_2008_STATUS_VALUES = Idna2008Status.values();
+  public static final IdnaMappingTable INSTANCE;
+  static {
```

Comment:
My goal is to put a small shim in front of this so that this class only gets loaded if your URL contains characters outside of `[A-Za-z0-9.-]`. Though I know it's tough to prevent classes from being loaded!